File: packages/common/src/feed.ts
import { Opaque, Maybe } from './helpers';
import { stringify, parse, Stringified } from './stringify';
import {
  FeedVersion,
  Address,
  Latitude,
  Longitude,
  City,
  NodeId,
  NodeCount,
  NodeDetails,
  NodeStats,
  NodeHardware,
  NodeLocation,
  BlockNumber,
  BlockHash,
  BlockDetails,
  Timestamp,
  Milliseconds,
  ChainLabel,
  AuthoritySetInfo,
} from './types';

export const Actions = {
  FeedVersion          : 0x00 as 0x00,
  BestBlock            : 0x01 as 0x01,
  BestFinalized        : 0x02 as 0x02,
  AddedNode            : 0x03 as 0x03,
  RemovedNode          : 0x04 as 0x04,
  LocatedNode          : 0x05 as 0x05,
  ImportedBlock        : 0x06 as 0x06,
  FinalizedBlock       : 0x07 as 0x07,
  NodeStats            : 0x08 as 0x08,
  NodeHardware         : 0x09 as 0x09,
  TimeSync             : 0x0A as 0x0A,
  AddedChain           : 0x0B as 0x0B,
  RemovedChain         : 0x0C as 0x0C,
  SubscribedTo         : 0x0D as 0x0D,
  UnsubscribedFrom     : 0x0E as 0x0E,
  Pong                 : 0x0F as 0x0F,
  AfgFinalized         : 0x10 as 0x10,
  AfgReceivedPrevote   : 0x11 as 0x11,
  AfgReceivedPrecommit : 0x12 as 0x12,
  AfgAuthoritySet      : 0x13 as 0x13,
  StaleNode            : 0x14 as 0x14,
};

export type Action = typeof Actions[keyof typeof Actions];
export type Payload = Message['payload'];

export namespace Variants {
  export interface MessageBase {
    action: Action;
  }

  export interface FeedVersionMessage extends MessageBase {
    action: typeof Actions.FeedVersion;
    payload: FeedVersion;
  }

  export interface BestBlockMessage extends MessageBase {
    action: typeof Actions.BestBlock;
    payload: [BlockNumber, Timestamp, Maybe<Milliseconds>];
  }

  export interface BestFinalizedBlockMessage extends MessageBase {
    action: typeof Actions.BestFinalized;
    payload: [BlockNumber, BlockHash];
  }

  export interface AddedNodeMessage extends MessageBase {
    action: typeof Actions.AddedNode;
    payload: [NodeId, NodeDetails, NodeStats, NodeHardware, BlockDetails, Maybe<NodeLocation>, Timestamp];
  }

  export interface RemovedNodeMessage extends MessageBase {
    action: typeof Actions.RemovedNode;
    payload: NodeId;
  }

  export interface LocatedNodeMessage extends MessageBase {
    action: typeof Actions.LocatedNode;
    payload: [NodeId, Latitude, Longitude, City];
  }

  export interface ImportedBlockMessage extends MessageBase {
    action: typeof Actions.ImportedBlock;
    payload: [NodeId, BlockDetails];
  }

  export interface FinalizedBlockMessage extends MessageBase {
    action: typeof Actions.FinalizedBlock;
    payload: [NodeId, BlockNumber, BlockHash];
  }

  export interface NodeStatsMessage extends MessageBase {
    action: typeof Actions.NodeStats;
    payload: [NodeId, NodeStats];
  }

  export interface NodeHardwareMessage extends MessageBase {
    action: typeof Actions.NodeHardware;
    payload: [NodeId, NodeHardware];
  }

  export interface TimeSyncMessage extends MessageBase {
    action: typeof Actions.TimeSync;
    payload: Timestamp;
  }

  export interface AddedChainMessage extends MessageBase {
    action: typeof Actions.AddedChain;
    payload: [ChainLabel, NodeCount];
  }

  export interface RemovedChainMessage extends MessageBase {
    action: typeof Actions.RemovedChain;
    payload: ChainLabel;
  }

  export interface SubscribedToMessage extends MessageBase {
    action: typeof Actions.SubscribedTo;
    payload: ChainLabel;
  }

  export interface UnsubscribedFromMessage extends MessageBase {
    action: typeof Actions.UnsubscribedFrom;
    payload: ChainLabel;
  }

  export interface PongMessage extends MessageBase {
    action: typeof Actions.Pong;
    payload: string; // just echo whatever `ping` sent
  }

  export interface AfgFinalizedMessage extends MessageBase {
    action: typeof Actions.AfgFinalized;
    payload: [Address, BlockNumber, BlockHash];
  }

  export interface AfgAuthoritySet extends MessageBase {
    action: typeof Actions.AfgAuthoritySet;
    payload: AuthoritySetInfo;
  }

  export interface AfgReceivedPrecommit extends MessageBase {
    action: typeof Actions.AfgReceivedPrecommit;
    payload: [Address, BlockNumber, BlockHash, Address];
  }

  export interface AfgReceivedPrevote extends MessageBase {
    action: typeof Actions.AfgReceivedPrevote;
    payload: [Address, BlockNumber, BlockHash, Address];
  }

  export interface StaleNodeMessage extends MessageBase {
    action: typeof Actions.StaleNode;
    payload: NodeId;
  }
}

export type Message =
  | Variants.FeedVersionMessage
  | Variants.BestBlockMessage
  | Variants.BestFinalizedBlockMessage
  | Variants.AddedNodeMessage
  | Variants.RemovedNodeMessage
  | Variants.LocatedNodeMessage
  | Variants.ImportedBlockMessage
  | Variants.FinalizedBlockMessage
  | Variants.NodeStatsMessage
  | Variants.NodeHardwareMessage
  | Variants.TimeSyncMessage
  | Variants.AddedChainMessage
  | Variants.RemovedChainMessage
  | Variants.SubscribedToMessage
  | Variants.UnsubscribedFromMessage
  | Variants.AfgFinalizedMessage
  | Variants.AfgReceivedPrevote
  | Variants.AfgReceivedPrecommit
  | Variants.AfgAuthoritySet
  | Variants.StaleNodeMessage
  | Variants.PongMessage;

/**
 * Data type to be sent to the feed. Passing through strings means we can only serialize once,
 * no matter how many feed clients are listening in.
 */
export interface SquashedMessages extends Array<Action | Payload> {};
export type Data = Stringified<SquashedMessages>;

/**
 * Serialize an array of `Message`s to a single JSON string.
 *
 * All messages are squashed into a single array of alternating opcodes and payloads.
 *
 * Action `string`s are converted to opcodes using the `actionToCode` mapping.
 */
export function serialize(messages: Array<Message>): Data {
  const squashed: SquashedMessages = new Array(messages.length * 2);
  let index = 0;

  messages.forEach((message) => {
    const { action, payload } = message;

    squashed[index++] = action;
    squashed[index++] = payload;
  })

  return stringify(squashed);
}

/**
 * Deserialize data to an array of `Message`s.
 */
export function deserialize(data: Data): Array<Message> {
  const json = parse(data);

  if (!Array.isArray(json) || json.length === 0 || json.length % 2 !== 0) {
    throw new Error('Invalid FeedMessage.Data');
  }

  const messages = new Array<Message>(json.length / 2);

  for (const index of messages.keys()) {
    const [ action, payload ] = json.slice(index * 2);

    messages[index] = { action, payload } as Message;
  }

  return messages;
}

File: backend/core/src/feed.rs
use serde::ser::{SerializeTuple, Serializer};
use serde::Serialize;
use std::mem;

use crate::node::Node;
use serde_json::to_writer;
use shared::types::{
    Address, BlockDetails, BlockHash, BlockNumber, NodeHardware, NodeIO, NodeId, NodeStats,
    Timestamp, NodeDetails,
};

pub mod connector;

use connector::Serialized;

pub trait FeedMessage {
    const ACTION: u8;
}

pub trait FeedMessageWrite: FeedMessage {
    fn write_to_feed(&self, ser: &mut FeedMessageSerializer);
}

impl<T> FeedMessageWrite for T
where
    T: FeedMessage + Serialize,
{
    fn write_to_feed(&self, ser: &mut FeedMessageSerializer) {
        ser.write(self)
    }
}

pub struct FeedMessageSerializer {
    /// Current buffer,
    buffer: Vec<u8>,
}

const BUFCAP: usize = 128;

impl FeedMessageSerializer {
    pub fn new() -> Self {
        Self {
            buffer: Vec::with_capacity(BUFCAP),
        }
    }

    pub fn push<Message>(&mut self, msg: Message)
    where
        Message: FeedMessageWrite,
    {
        let glue = match self.buffer.len() {
            0 => b'[',
            _ => b',',
        };

        self.buffer.push(glue);
        self.write(&Message::ACTION);
        self.buffer.push(b',');
        msg.write_to_feed(self);
    }

    fn write<S>(&mut self, value: &S)
    where
        S: Serialize,
    {
        let _ = to_writer(&mut self.buffer, value);
    }

    pub fn finalize(&mut self) -> Option<Serialized> {
        if self.buffer.is_empty() {
            return None;
        }

        self.buffer.push(b']');

        let bytes = mem::replace(&mut self.buffer, Vec::with_capacity(BUFCAP)).into();

        Some(Serialized(bytes))
    }
}

macro_rules! actions {
    ($($action:literal: $t:ty,)*) => {
        $(
            impl FeedMessage for $t {
                const ACTION: u8 = $action;
            }
        )*
    }
}

actions! {
    0x00: Version,
    0x01: BestBlock,
    0x02: BestFinalized,
    0x03: AddedNode<'_>,
    0x04: RemovedNode,
    0x05: LocatedNode<'_>,
    0x06: ImportedBlock<'_>,
    0x07: FinalizedBlock,
    0x08: NodeStatsUpdate<'_>,
    0x09: Hardware<'_>,
    0x0A: TimeSync,
    0x0B: AddedChain<'_>,
    0x0C: RemovedChain<'_>,
    0x0D: SubscribedTo<'_>,
    0x0E: UnsubscribedFrom<'_>,
    0x0F: Pong<'_>,
    0x10: AfgFinalized,
    0x11: AfgReceivedPrevote,
    0x12: AfgReceivedPrecommit,
    0x13: AfgAuthoritySet,
    0x14: StaleNode,
    0x15: NodeIOUpdate<'_>,
}

#[derive(Serialize)]
pub struct Version(pub usize);

#[derive(Serialize)]
pub struct BestBlock(pub BlockNumber, pub Timestamp, pub Option<u64>);

#[derive(Serialize)]
pub struct BestFinalized(pub BlockNumber, pub BlockHash);

pub struct AddedNode<'a>(pub NodeId, pub &'a Node);

#[derive(Serialize)]
pub struct RemovedNode(pub NodeId);

#[derive(Serialize)]
pub struct LocatedNode<'a>(pub NodeId, pub f32, pub f32, pub &'a str);

#[derive(Serialize)]
pub struct ImportedBlock<'a>(pub NodeId, pub &'a BlockDetails);

#[derive(Serialize)]
pub struct FinalizedBlock(pub NodeId, pub BlockNumber, pub BlockHash);

#[derive(Serialize)]
pub struct NodeStatsUpdate<'a>(pub NodeId, pub &'a NodeStats);

#[derive(Serialize)]
pub struct NodeIOUpdate<'a>(pub NodeId, pub &'a NodeIO);

#[derive(Serialize)]
pub struct Hardware<'a>(pub NodeId, pub &'a NodeHardware);

#[derive(Serialize)]
pub struct TimeSync(pub u64);

#[derive(Serialize)]
pub struct AddedChain<'a>(pub &'a str, pub usize);

#[derive(Serialize)]
pub struct RemovedChain<'a>(pub &'a str);

#[derive(Serialize)]
pub struct SubscribedTo<'a>(pub &'a str);

#[derive(Serialize)]
pub struct UnsubscribedFrom<'a>(pub &'a str);

#[derive(Serialize)]
pub struct Pong<'a>(pub &'a str);

#[derive(Serialize)]
pub struct AfgFinalized(pub Address, pub BlockNumber, pub BlockHash);

#[derive(Serialize)]
pub struct AfgReceivedPrevote(
    pub Address,
    pub BlockNumber,
    pub BlockHash,
    pub Option<Address>,
);

#[derive(Serialize)]
pub struct AfgReceivedPrecommit(
    pub Address,
    pub BlockNumber,
    pub BlockHash,
    pub Option<Address>,
);

#[derive(Serialize)]
pub struct AfgAuthoritySet(
    pub Address,
    pub Address,
    pub Address,
    pub BlockNumber,
    pub BlockHash,
);

#[derive(Serialize)]
pub struct StaleNode(pub NodeId);

impl FeedMessageWrite for AddedNode<'_> {
    fn write_to_feed(&self, ser: &mut FeedMessageSerializer) {
        let AddedNode(nid, node) = self;

        let details = node.details();
        let details = (
            &details.name,
            &details.implementation,
            &details.version,
            &details.validator,
            &details.network_id,
        );

        ser.write(&(
            nid,
            details,
            node.stats(),
            node.io(),
            node.hardware(),
            node.block_details(),
            &node.location(),
            &node.startup_time(),
        ));
    }
}
